[GAWB-2384] Notebook Service: Scala code to proxy websocket connections

rtitle opened the pull request:
https://broadinstitute.atlassian.net/browse/GAWB-2384

Still not done:
- Test on a real cluster created by the Leo PUT endpoint (dependent on [GAWB-2268](https://broadinstitute.atlassian.net/browse/GAWB-2268))
- Test the Leo proxy behind the Orch proxy
   - I suspect the Orch proxy may need to be tweaked slightly to pass through websocket connections
- Implement SSL connections to Jupyter
   - Will require some changes in the Jupyter server config

----

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

rtitle commented on docker/jupyter_notebook_config.py:
```diff
@@ -0,0 +1,17 @@
+c.NotebookApp.ip = '*'
+c.NotebookApp.port = 8000
+c.NotebookApp.open_browser = False
+c.NotebookApp.base_url = '/notebooks/test/'
```
So, I think for the proxy to work correctly, we'll want to set `c.NotebookApp.base_url` and `c.NotebookApp.webapp_settings` for real notebook servers as well. Hopefully that will be possible through the init script.

helgridly replied:
I haven't seen notebooks-land in a while, but IIRC these options let you set the notebooks landing page to `host.com/things_you_specify/tree` rather than just `host.com/tree`. Why does the proxy care about this, beyond having to know what it is?

rtitle replied:
So we want our proxy to live at the path:
```
http://leo:8080/notebooks/{id}
```
It gets immensely easier if the Notebook server is listening on the same path. Otherwise the client-side notebook code tries to load assets at the root path (http://leo:8080/) which does not get proxied.

I admit I don't fully understand all the calls being made, but I had problems with other configurations. Things I could try:
1. Look into exactly what `NotebookApp.base_url` is doing in Jupyter code. Is there some way we can simulate this in our reverse proxy, and avoid changing `jupyter_notebook_config.py`?
2. Make the Leo proxy listen at / as well as /notebooks/id. The root path might just be needed for static resources. Need to make sure it doesn't clash with Swagger which also listens at /.

helgridly replied:
Ah, I see. That being the case I'm okay with this. It doesn't really affect the user's experience, since all they see is the Leo URL anyway.

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+    // TODO eventually we will lookup the outgoing address from the notebook ID
+    val flow = Http(system).outgoingConnection(request.uri.authority.host.address, jupyterPort)
+    val newHeaders = filterHeaders(request.headers)
+    val newUri = Uri(path = request.uri.path, queryString = request.uri.queryString())
```
Not exactly sure why, but I need to strip out the `Uri.Authority` in order for the proxy to work. Hence building `newUri` and `newRequest`.

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+    val handler = Source.single(newRequest)
+      .via(flow)
+      .runWith(Sink.head)
+      .flatMap(_.toStrict(5 seconds))
```
Also it only works with Jupyter if I call `toStrict`. My unit tests work without it, so Jupyter might be doing something strange. `toStrict` means fully loading the response entity into memory before returning it, as opposed to streaming the data. I hope this won't be an issue for us, as Jupyter http assets seem to be pretty small (it loads some html/js/css files but that's about it).

helgridly replied:
i'm pretty sure that large images or datatables sent back to the user from the python kernel go over the websocket, but please doublecheck this. because those can get biiiiiig.

rtitle replied:
Will double check, but I think everything from the kernel is over the websocket.

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+      }
+    }
+
+  private def handleHttpRequest(request: HttpRequest): Future[HttpResponse] = {
```
Basis for this method stolen from https://gist.github.com/rklaehn/3aa3215046df2c0a7795

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+    handler
+  }
+
+  private def handleWebSocketRequest(request: HttpRequest, upgrade: UpgradeToWebSocket): Future[HttpResponse] = {
```
Basis for this method stolen from https://github.com/akka/akka-http/issues/1289#issuecomment-316269886

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+
+  private def handleWebSocketRequest(request: HttpRequest, upgrade: UpgradeToWebSocket): Future[HttpResponse] = {
+    logger.info(s"Opening websocket connection to ${request.uri.authority.host.address}")
+    val flow = Flow.fromSinkAndSourceMat(Sink.asPublisher[Message](fanout = false), Source.asSubscriber[Message])(Keep.both)
```
The websocket proxy doesn't have to be `Strict` at least! (there's a `Message.Strict` class for doing that)

jmthibault79 replied:
Could you add some comments about what `Publisher`, `Subscriber`, and `Keep` are doing here?

rtitle replied:
Yes

helgridly replied:
+1. I think an explanation of this stuff would be a great cross-team tech talk topic!

rtitle commented on src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutesSpec.scala:
```diff
@@ -0,0 +1,183 @@
+    // Flow to hit the proxy server
+    val webSocketFlow = Http().webSocketClientFlow(WebSocketRequest(Uri("ws://localhost:9000/notebooks/websocket"))).map {
+      case m: TextMessage.Strict => m.text
+      case _ => throw new IllegalArgumentException("ProxyRoutesSpec only supports strict messages")
```
Maybe should add a case for non-strict messages..

jmthibault79 commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+    pathPrefix(Segment) { notebookId =>  // TODO notebookId not currently used
+      extractRequest { request =>
+        complete {
+          request.header[UpgradeToWebSocket] match {
```
What determines whether we see this header, and what is different about the action we should take?

rtitle replied:
I think it's an akka-http thing -- if akka-http receives a websocket request (e.g. `wss://foo/bar`) then it will inject this "virtual" header which has methods for handling the websocket traffic. See: https://github.com/akka/akka-http/blob/master/akka-http-core/src/main/scala/akka/http/scaladsl/model/ws/UpgradeToWebSocket.scala

So basically it distinguishes websocket from non-websocket traffic.

jmthibault79 replied:
Ah!  I had thought this was for websocket only.  ok

jmthibault79 commented on src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutesSpec.scala:
```diff
@@ -0,0 +1,186 @@
+    pathPrefix("notebooks") {
+      extractRequest { request =>
+        path("websocket") {
+          println("in a websocet")
```
rm

rtitle replied:
oops, I can't even spell my debug statements correctly

jmthibault79 commented on src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutesSpec.scala:
```diff
@@ -0,0 +1,186 @@
+
+  "ProxyRoutes" should "listen on /notebooks/{id}/..." in {
+    Get("/notebooks/test").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      handled shouldBe true
```
ooh we need to use this everywhere

jmthibault79 commented on src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutesSpec.scala:
```diff
@@ -0,0 +1,186 @@
+      case _ => throw new IllegalArgumentException("ProxyRoutesSpec only supports strict messages")
+    }
+
+    // Glue together the source, sink, and flow. Returns a tuple of:
```
Could you elaborate?  I'm not following this at all.

rtitle replied:
I will try to add more explanation to `ProxyRoutes` and `ProxyRoutesSpec` (assuming I understand it myself). This akka-stream stuff is hard to follow.

jmthibault79 replied:
Thanks!

And thanks for stream-pioneering more generally.

helgridly commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+  val jupyterPort = 8000
+
+  val proxyRoutes: Route =
+    pathPrefix(Segment) { notebookId =>  // TODO notebookId not currently used
```
this is going to look like `GET /api/notebooks/<google_project>/<cluster_name>`, fwiw

rtitle replied:
ok

helgridly commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+        complete {
+          request.header[UpgradeToWebSocket] match {
+            case Some(upgrade) => handleWebSocketRequest(request, upgrade)
+            case None => handleHttpRequest(request)
```
please break these functions out into a `ProxyService`. see `LeoRoutes` -> `LeonardoService` for pattern

helgridly replied:
to which you'll need to pass the notebook details too

rtitle replied:
ok, makes sense

helgridly commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+
+  private def handleHttpRequest(request: HttpRequest): Future[HttpResponse] = {
+    logger.info(s"Opening http connection to ${request.uri.authority.host.address}")
+    // TODO eventually we will lookup the outgoing address from the notebook ID
```
so as written here, it proxies back to wherever the request came from, right? and when we do it for reals, the `ProxyService` will hit the `ClusterComponent` and off we go.

rtitle replied:
a) yes it forwards the requests to the notebook server and returns the responses back to the caller.
b) I wonder if it will introduce too much latency to hit `ClusterComponent` for _every_ proxied notebook request. Maybe we could have a cache of `(google_project, cluster_name) -> IP` which the proxy could look up, and only query the DB on cache misses.

helgridly replied:
I'm assuming there's a query cache somewhere around slick / the db anyway. We'll load test it; this should be an easy thing to add later.

helgridly commented on src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutesSpec.scala:
```diff
@@ -15,9 +15,7 @@ import scala.concurrent.ExecutionContext
 
-  class TestLeoRoutes(leonardoService: LeonardoService)
-                     (override implicit val materializer: Materializer, override val executionContext: ExecutionContext)
-    extends LeoRoutes(leonardoService, SwaggerConfig())(materializer, executionContext)
```
this went away because `LeoRoutes` is an `ActorSystem` now?

rtitle replied:
It didn't go away, I just removed the unnecessary implicits in the `TestLeoRoutes` constrcutor :)

rtitle replied:
(`LeoRoutes` is not an `ActorSystem`)

helgridly commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutes.scala:
```diff
@@ -0,0 +1,93 @@
+  * Created by rtitle on 8/4/17.
+  */
+trait ProxyRoutes { self: LazyLogging =>
+  implicit val system: ActorSystem
```
Why does this need to be an `ActorSystem`? Because of the flow magic?

rtitle replied:
Looks like the `ActorSystem` is needed for making the outgoing HTTP requests. This is just an abstract method in the trait -- the ActorSystem gets passed into the `LeoRoutes` constructor.

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala:
```diff
@@ -48,11 +46,10 @@ class LeoRoutes(val leonardoService: LeonardoService, val swaggerConfig: Swagger
     swaggerRoutes ~
-      pathPrefix("api") { leoRoutes }
+    pathPrefix("api") { leoRoutes } ~
+    pathPrefix("notebooks") { proxyRoutes }
```
👍

rtitle commented:
Yay! I'll make the requested changes.

rtitle commented:
@jmthibault79 @helgridly this is ready for re-review. Changes since you last looked:
1. Moved most of the proxy logic out of `ProxyRoutes` and into a `ProxyService` class.
2. Integrated with the Slick code, and actually lookup the Notebook IP from the database given a google project and cluster name.
3. Added a lot more comments to `ProxyService`

Thanks!

helgridly commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/ProxyService.scala:
```diff
@@ -0,0 +1,154 @@
+  */
+class ProxyService(dbRef: DbReference)(implicit val system: ActorSystem, materializer: ActorMaterializer, executionContext: ExecutionContext) extends LazyLogging {
+  // comes from jupyter_notebook_config.py
+  private final val JupyterPort = 8000
```
conf this

rtitle replied:
ok

rtitle replied:
done
